dlc/src/digit_decomposition.rs

```diff
@@ -0,0 +1,673 @@
+//! Utility functions to decompose numeric outcome values
+
+use super::{Error, Payout, RangeOutcome};
+
+struct PrefixRange {
+    start: Vec<usize>,
+    end: Vec<usize>,
+    prefix: Vec<usize>,
+}
+
+/// Decompose a numeric value into digits in the specified base. If the decomposed
+/// value contains less than `nb_digits`, zeroes will be prepended to reach `nb_digits`
+/// size.
+pub fn decompose_value(mut value: usize, base: usize, nb_digits: usize) -> Vec<usize> {
+    let mut res = Vec::new();
+
+    while value > 0 {
+        res.push(value % base);
+        value = ((value as f64) / (base as f64)).floor() as usize;
+    }
+
+    while res.len() < nb_digits {
+        res.push(0);
+    }
+
+    res.into_iter().rev().collect()
+}
+
+/// Takes a decomposed representation of a numerical value in a given base and returns
+/// the represented value as a `usize`
+pub fn compose_value(values: &[usize], base: usize) -> usize {
+    let mut composed = 0;
+    for i in 0..values.len() {
+        let pow = values.len() - i - 1;
+        composed += values[i] * base.pow(pow as u32);
+    }
+
+    composed
+}
+
+/// Takes a vector or `RangeOutcome` and (if necessary) adds one at the beginning
+/// and at the end of it to cover the full range of possible outcome values
+pub fn get_max_ranges(
+    original_outcomes: Vec<RangeOutcome>,
+    base: usize,
+    nb_digits: usize,
+) -> Vec<RangeOutcome> {
+    let mut outcomes: Vec<RangeOutcome> = original_outcomes;
+    if outcomes[0].start != 0 {
+        outcomes[0] = RangeOutcome {
+            start: 0,
+            count: outcomes[0].count + outcomes[0].start,
+            payout: Payout {
+                offer: outcomes[0].payout.offer,
+                accept: outcomes[0].payout.accept,
+            },
+        };
+    }
+    let last_index = outcomes.len() - 1;
+    let last_outcome = &outcomes[last_index];
+    let max_value = base.pow(nb_digits as u32);
+    if last_outcome.start + last_outcome.count != max_value {
+        outcomes[last_index] = RangeOutcome {
+            start: last_outcome.start,
+            count: max_value - last_outcome.start,
+            payout: Payout {
+                offer: last_outcome.payout.offer,
+                accept: last_outcome.payout.accept,
+            },
+        }
+    }
+
+    outcomes
+}
+
+fn separate_prefix(
```

Comment:
nit: I think these functions are pretty straightforward but for someone reading this who hasn't yet looked at the spec it might be nice to have cursory doc comments on this function and the ones below

Reply:
Added some doc hope it helps